jupyterlab/jupyterlab
# Update notebook kernel selector

The "change kernel" dialog is complex, considering that it is primarily just a `<select>` dropdown. Since the release of the running kernels tab, though, the notion of what we mean when we say "kernel" has become somewhat muddled. The "change kernel" dialog actually shows you kernel specs against which a new kernel instance can be created or running **sessions** in JupyterLab. However, the running kernels sidebar shows you a list of running **kernels** and uses any session that has more information about a kernel instance to update the UI, but not to group the kernels.

This PR updates the "change kernel" UI:
- It slightly changes the language in the `<optgroup>` sections of the dropdown to clarify that "preferred" means "using the programming language I used last time".
- It lists every running kernel as a choice – even if there is no JupyterLab session associated with it.

## References
Fixes https://github.com/jupyterlab/jupyterlab/issues/15255
cf. https://github.com/jupyterlab/jupyterlab/issues/16340

## Code changes
- This PR updates the change kernel dialog to include all kernel sessions as well as all kernelspecs.
- It adds an optional kernel manager to session contexts, in future versions, this should be required.
- It exposes a new function `SessionContextDialogs.kernelOptions()` that returns grouped kernelspecs and running kernel options that can be used to build a UI (*e.g.*, a `<select>` tag or React component).
- It adds a test for checking whether external kernels that JupyterLab's REST endpoint are reflected in the kernel options presented to the user.

## User-facing changes

#### before
<img width="908" alt="before" src="https://github.com/jupyterlab/jupyterlab/assets/159529/c2b77790-453b-475d-9a91-2f00502df355">

#### after
<img width="926" alt="after" src="https://github.com/jupyterlab/jupyterlab/assets/159529/7c55c8d4-9f10-412f-83ed-cd1c3e3f4e37">

## Backwards-incompatible changes

N/A

**NB** The `kernelManager` option for `SessionContext` is currently optional. In the next major version of this package, we should make it required.

The `SessionContext` instances created by JupyterLab core have all been updated to pass in a kernel manager.

If there is no kernel manager available (i.e. if there is a plugin that creates `SessionContext` instances that do not have a kernel manager), the dropdown will harvest all known kernels from running sessions, which is roughly the old behavior.

## To test this yourself

I have updated the tests to include this use case, but I can show how to test this pretty easily in your own environment, it doesn't require installing anything exotic if all we want is to see these UI changes and don't care about whether there is a real underlying kernel to work with.

Create a `kernel.json` file with contents like this in an otherwise empty directory (I call mine `external_kernels`):

```json
{
    "transport": "tcp",
    "ip": "127.0.0.1",
    "control_port": 63607,
    "shell_port": 49458,
    "stdin_port": 54582,
    "iopub_port": 50266,
    "hb_port": 59840,
    "signature_scheme": "hmac-sha256",
    "key": "883965661293433ea9c07538beeb8451",
    "kernel_name": "mock-kernel"
}
```

Start JupyterLab with `ServerApp.allow_external_kernels` set to `True` and `ServerApp.external_connection_dir` set to the directory you created above. The command I used also included `--dev-mode` and `--watch` and a blank identity provider because I was in development mode as well as `extensions-in-dev-mode` to allow me to see `jupyterlab_launchpad`:

```
jupyter lab \
--dev-mode --extensions-in-dev-mode --watch \
--no-browser \
--IdentityProvider.token="" \
--ServerApp.external_connection_dir=./external_kernels \
--ServerApp.allow_external_kernels=True
```
